src/storage/core.rs

```diff
@@ -416,53 +418,93 @@ where
     /// # Errors
     /// Fails after any disk IO errors.
     pub async fn read_all(&self, key: impl AsRef<K>) -> Result<Vec<Entry>> {
+        let mut entries = self.read_all_with_deletion_marker(key).await?;
+        if let Some(e) = entries.last() {
+            if e.header().is_deleted() {
+                entries.truncate(entries.len() - 1);
+            }
+        }
+        Ok(entries)
+    }
+
+    /// Returns entries with matching key and deletion marker, if any
+    /// # Errors
+    /// Fails after any disk IO errors.
+    pub async fn read_all_with_deletion_marker(&self, key: impl AsRef<K>) -> Result<Vec<Entry>> {
         let key = key.as_ref();
         let mut all_entries = Vec::new();
         let safe = self.inner.safe.read().await;
         let active_blob = safe
             .active_blob
             .as_ref()
             .ok_or_else(Error::active_blob_not_set)?;
-        if let Some(entries) = active_blob.read().await.read_all_entries(key).await? {
-            debug!(
-                "storage core read all active blob entries {}",
-                entries.len()
-            );
-            all_entries.extend(entries);
+        let entries = active_blob
+            .read()
+            .await
+            .read_all_entries_with_deletion_marker(key)
+            .await?;
+        debug!(
+            "storage core read all active blob entries {}",
+            entries.len()
+        );
+        if let Some(e) = entries.last() {
+            if e.header().is_deleted() {
+                return Ok(entries);
+            }
         }
+        all_entries.extend(entries);
         let blobs = safe.blobs.read().await;
-        let entries_closed_blobs = blobs
-            .iter_possible_childs(key)
-            .map(|b| b.1.data.read_all_entries(key))
-            .collect::<FuturesUnordered<_>>();
-        entries_closed_blobs
-            .try_filter_map(future::ok)
-            .try_for_each(|v| {
-                debug!("storage core read all closed blob {} entries", v.len());
-                all_entries.extend(v);
-                future::ok(())
-            })
-            .await?;
-        debug!("storage core read all total {} entries", all_entries.len());
+        let mut futures = blobs
+            .iter_possible_childs_rev(key)
+            .map(|b| b.1.data.read_all_entries_with_deletion_marker(key))
+            .collect::<FuturesOrdered<_>>();
+        while let Some(data) = futures.next().await {
+            let entries = data?;
+            if let Some(e) = entries.last() {
+                if e.header().is_deleted() {
+                    all_entries.extend(entries);
```

Comment:
`extend` is performed regardless of whether the deletion marker is present or not. So I think it would be good to move this line up to the position right after the `entries` are assigned (`let entries = data?;`)

Reply:
Extend requires ownership in this case, and if we add it earlier we need to check last record in `all_entries` and remember if any was added from `entries` this time, which will require new flag. I think this way there are less lines of code involved